src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -496,17 +520,20 @@
 (defn- flow-field-metadata
   "Merge information about fields from `source-metadata` into the returned `cols`."
   [source-metadata cols]
-  (let [field-id->metadata (u/key-by :id source-metadata)]
+  (let [ref->metadata (u/key-by (comp u/field-ref->key :field_ref) source-metadata)]
```

Comment:
Previously merging has been keyed off of `:id` which doesn't exist always and starts hanging metadata on all columns without an id (all native columns, all expressions, etc). I figure by field_ref is the best. And You can see from the source of this function that it only takes the `[:field <id-or-name>]` part and omits the map at the end that can vary with type info, join alias information, etc.